allennlp/common/file_utils.py

```diff
@@ -144,3 +149,95 @@ def get_from_cache(url: str, cache_dir: str = None) -> str:
+    """ Returns the file extension, by default case-lowered and including the dot. """
+    ext = os.path.splitext(path)[1]
+    ext = ext if dot else ext[1:]
+    return ext.lower() if lower else ext
+
+
+class CompressedFileUtils:
+
+    SUPPORTED_FORMATS = {'.gz', '.zip'}
+    READ_MODE_CHOICES = {'t', 'rt', 'b', 'rb'}
+    DEFAULT_ENCODING = 'utf-8'
+
+    @staticmethod
+    def read(path: str, mode: str = 't', encoding: str = None,
+             file_format: str = None):
+        """
+        Open an eventually compressed file in binary or text mode (default: text mode
+        with utf-8 encoding). The currently supported compressed file formats are: gzip, zip.
+
+        Parameters
+        ----------
+        path: str
+            Path to the file to read
+        mode: str
+            Reading mode; it can be 't' or 'rt' for text mode, 'b' or 'rb' for binary mode.
+        encoding: str
+            Text encoding of the text file. This must be left to None when reading in binary mode
+            or an exception will be raised.
+        file_format: str
+            If ``file_format == None``, the format is inferred from the extension.
+            If the file format (specified or inferred) is not supported, an exception is raised.
+
+        Returns
+        -------
+        When used in text mode, it returns a :class:`io.TextIOWrapper`.
+        When used in binary mode, the return type depends on the specific input file format.
+        """
+        read_mode_choices = CompressedFileUtils.READ_MODE_CHOICES
+        if mode not in read_mode_choices:
+            raise ValueError("Invalid mode: {}. Supported modes are: {}"
+                             .format(mode, read_mode_choices))
+
+        file_format = file_format or get_file_extension(path)
+
+        if file_format == ".gz":
+            logger.info("Opening gzipped file: %s", path)
+            binary_reader = gzip.open(path, "rb")
+
+        elif file_format == ".zip":
+            logger.info("Opening zipped file: %s", path)
+            zfile = zipfile.ZipFile(path, "r")
+            assert len(zfile.namelist()) == 1, "Multiple files are contained in the zip archive " + path
+            filename = zfile.namelist()[0]
+            binary_reader = zfile.open(filename, "r")
+
+        else:
+            raise ValueError('Unsupported file format: {}. Supported formats are: {}'
+                             .format(file_format, CompressedFileUtils.SUPPORTED_FORMATS))
+
+        if mode == 't':
+            encoding = encoding or CompressedFileUtils.DEFAULT_ENCODING
+            return io.TextIOWrapper(binary_reader, encoding=encoding)
+        else:
+            if encoding is not None:    # same behavior of built-in open()
+                raise ValueError("Binary mode doesn't take an encoding argument")
+            return binary_reader
+
+
+def read_maybe_compressed_file(path: str, mode: str = 't', encoding: str = None,
```

Comment:
same comment, I don't think this function should be called `read`, in one of its cases it directly calls `open`